Added statistics function

ByteHamster opened the pull request:
#1743

mfietz commented on app/src/main/res/layout/statistics_fragment.xml:
```diff
@@ -0,0 +1,35 @@
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:orientation="vertical"
+    android:padding="10dp">
```
10dp is a rather unusual value
You might want to take a look at https://www.google.com/design/spec/layout/metrics-keylines.html#metrics-keylines-keylines-spacing

mfietz commented on app/src/main/java/de/danoeh/antennapod/adapter/StatisticsListAdapter.java:
```diff
@@ -0,0 +1,105 @@
+
+    @Override
+    public long getItemId(int position) {
+        return 0;
```
you should return something useful here, e.g. ``feed.getId()``
See https://developer.android.com/intl/es/reference/android/support/v7/widget/RecyclerView.Adapter.html#setHasStableIds(boolean)

mfietz commented on app/src/main/java/de/danoeh/antennapod/adapter/StatisticsListAdapter.java:
```diff
@@ -0,0 +1,105 @@
+ */
+public class StatisticsListAdapter extends BaseAdapter {
+    private Context context;
+    ArrayList<DBReader.StatisticsItem> feedTime = new ArrayList<>();
```
Effective Java, item 52: Refer to objects by their interfaces

mfietz commented on app/src/main/java/de/danoeh/antennapod/adapter/StatisticsListAdapter.java:
```diff
@@ -0,0 +1,105 @@
+            LayoutInflater inflater = (LayoutInflater) context
+                    .getSystemService(Context.LAYOUT_INFLATER_SERVICE);
+
+            convertView = inflater.inflate(R.layout.nav_feedlistitem, parent, false);
```
While this creative re-use of a layout, I'm not sure this is a good idea. You create an additional, non-obvious dependency here. Whenever we change the navbar listitem layout in the future, we would also have to consider that the layout is also used here.
You are also not using failure at all and using the former count (of new/unused/unused+new episodes) for the total runtime also has a rather bad smell to it. You should at least give your variables better names!

mfietz commented on app/src/main/java/de/danoeh/antennapod/adapter/StatisticsListAdapter.java:
```diff
@@ -0,0 +1,105 @@
+
+    private String timeToText(long time) {
+        int hours = (int) time / 3600;
+        int remainder = (int) time - hours * 3600;
```
``time % 3600`` ?

mfietz commented on app/src/main/java/de/danoeh/antennapod/adapter/StatisticsListAdapter.java:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    private String timeToText(long time) {
+        int hours = (int) time / 3600;
```
might even be nicer to use [TimeUnit](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/TimeUnit.html) here

ByteHamster replied:
I prefer a conversion that is done automatically: The method could also return days if the number of hours gets too high without requiring additional code at other places

mfietz replied:
Must be a misunderstanding. Just saying that 
``long hours = TimeUnit.HOURS.convert(time, TimeUnit.MILLISECONDS);``
is more readable than just 
``int hours = (int) time / 3600;``

mfietz commented on app/src/main/java/de/danoeh/antennapod/fragment/StatisticsFragment.java:
```diff
@@ -0,0 +1,106 @@
+    public void onStart() {
+        super.onStart();
+        totalTimeTextView.setText("...");
+        feedStatisticsList.setVisibility(View.GONE);
```
Not sure which of the options is better, but you see these kind of declarations quite often already in the layout file

ByteHamster replied:
The reason here is that onStart is also called when resuming the fragment. Gathering statistics might take a second so it should clear previously added data

mfietz replied:
Should set the text and change the visibility in ``loadStats()`` then to make the connection more obvious

mfietz commented on core/src/main/res/values-de/strings.xml:
```diff
@@ -28,6 +29,9 @@
   <string name="recently_published_episodes_label">Zuletzt veröffentlicht</string>
   <string name="episode_filter_label">Nur neue Episoden anzeigen</string>
+  <!-- Statistics fragment -->
+  <string name="total_time_listened_to_podcasts">Gesamtzeit gespielter Podcasts:</string>
```
This is how translations work. We push the default ``strings.xml`` to transifex. People translate the individual strings; we later pull these. That means that changes to all ``strings.xml`` files except the default one are overwritten once we pull from transifex.

ByteHamster replied:
But it does not break something if I keep the translations here? Or do I need to delete them again?

mfietz replied:
Does not break anything, no. Just superfluous.

TomHennen replied:
It actually may cause a problem when pulling down new translations.  Please remove this change from the PR.

mfietz commented on core/src/main/java/de/danoeh/antennapod/core/storage/DBReader.java:
```diff
@@ -918,6 +918,84 @@ public static FeedMedia getFeedMedia(final long mediaId) {
+            List<FeedItem> items = getFeed(feed.getId()).getItems();
+            for(FeedItem item : items) {
+                FeedMedia media = item.getMedia();
+                if(item.isPlayed()) {
```
check ``media != null`` here (NullPointerException)

mfietz commented on app/src/main/res/layout/statistics_fragment.xml:
```diff
@@ -0,0 +1,35 @@
+    <TextView
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:textAppearance="?android:attr/textAppearanceLarge"
```
You set the text appearance to large just to overwrite the size 3 lines later? o_O

mfietz commented on core/src/main/res/values/strings.xml:
```diff
@@ -33,6 +34,10 @@
     <string name="episode_filter_label">Show only new Episodes</string>
 
+    <!-- Statistics fragment -->
+    <string name="total_time_listened_to_podcasts">Total time listened to podcasts:</string>
```
Nitpick:
In fact, this might not be true. I listen to podcasts at a speed of 1.75x. I only listened about ``totalTime/1.75`` hours...
(But I can't come up with a better text at the top of my head, either)

ByteHamster replied:
Total time of podcasts played?

mfietz commented:
This could also use a progress indicator. If you have a lot of podcasts (50+), there is a noticeable delay before the statistics are shown.

ByteHamster commented:
Where should I put the method `timeToText`? Currently I have added the method twice.

mfietz commented:
``de.danoeh.antennapod.core.util.Converter``
You can probably even use ``getDurationStringShort()`` instead of your own implementation

mfietz commented on app/src/main/java/de/danoeh/antennapod/fragment/StatisticsFragment.java:
```diff
@@ -84,8 +90,7 @@ private void loadStats() {
         int hours = (int) time / 3600;
-        int remainder = (int) time - hours * 3600;
-        int mins = remainder / 60;
+        int mins = (int) time % 3600;
```
this only works for the remainder.
still have to divide by 60

ByteHamster replied:
Oh sorry... I added my own method to `Converter` because it contains the word "hours". Appending it manually every time is so much additional code

ByteHamster commented:
Shall I squash the commits now do they can be merged?

TomHennen commented:
A couple of things:

1. I really don't like the access in the Nav bar. Please move it in to 'Settings'. I know it's not a 'setting' per se, but I don't think this will be frequently accessed. Settings makes the most sense.

2. It says 00:22 hours.  Does that seem right? It's confusing as to if it's saying '22 hours' or '0 hours 22 minutes'. (I know it's the latter, but it may confuse others) The only time using hours would make sense in my mind is if you were reporting fractions of an hour and not minutes.  E.g.  '0.4 hours'

ByteHamster commented:
Removed strings, moved to settings, changed hours to fractions.

I squashed all together to a single commit and hope that it's now ready to merge :)

TomHennen commented:
Thanks for this contribution @ByteHamster!
